File: depend/libcmb/perl/tests/test1.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cmb;

my $config = {
	debug => 0,
	nul_terminate => 0,
	show_empty => 0,
	show_numbers => 1,
	delimiter => ",",
	prefix => "\t\t[",
	suffix => "]",
	size_min => 2,
	size_max => 3,
	count => 6,
	start => 4,
};
print "my \$config = {\n";
print join "", map { qq/\t$_ => "$config->{$_}",\n/ } sort keys %$config;
print "};\n";


my $vers = 1; # Long version
my @items = qw/a b c d/;
my $count = $#items + 1;
my $ilist = join ", ", @items;
my $seq = 1;
my $res;

#
# Method A:
#
printf "==> Method A:\n";
my $cmbA = new Cmb $config;
printf "\tnew Cmb created\n";
printf "\tCmb::cmb_version(): %s\n", Cmb::cmb_version();
printf "\tCmb::cmb_version(%i): %s\n", $vers, Cmb::cmb_version($vers);
printf "\tsize_min=%u size_max=%u\n", $config->{size_min}, $config->{size_max};
printf "\tCmb::cmb_count(config, %u) = %u\n", $count,
	Cmb::cmb_count($cmbA, $count);
printf "\tCmb::print(config, %u, %u, [%s]):\n", $seq, $count, $ilist;
$res = Cmb::print($cmbA, $seq, $count, \@items);
printf "\t\tRESULT: %i\n", $res;
printf "\tCmb::cmb(config, %u, [%s]):\n", $count, $ilist;
printf "\t\tNOTE: { start => %u, count => %u }\n",
	$config->{start}, $config->{count};
$res = Cmb::cmb($cmbA, $count, \@items);
printf "\t\tRESULT: %i\n", $res;

#
# Method B:
#
printf "==> Method B:\n";
my $cmbB = new Cmb;
printf "\tnew Cmb created\n";
$cmbB->config($config);
printf "\t\$cmb->version(): %s\n", $cmbB->version();
printf "\t\$cmb->version(%i): %s\n", $vers, $cmbB->version($vers);
printf "\tsize_min=%u size_max=%u\n", $config->{size_min}, $config->{size_max};
printf "\t\$cmb->count(%u) = %u\n", $count, $cmbB->count($count);
printf "\t\$cmb->print(%u, %u, [%s]):\n", $seq, $count, $ilist;
$res = $cmbB->print($seq, $count, \@items);
printf "\t\tRESULT: %i\n", $res;
printf "\t\$cmb->cmb(%u, [%s]):\n", $count, $ilist;
printf "\t\tNOTE: start=%u count=%u\n", $config->{start}, $config->{count};
$res = $cmbB->cmb($count, \@items);
printf "\t\tRESULT: %i\n", $res;

#
# Callback Method:
#
printf "==> Callback Method:\n";
my $cmbC = new Cmb $config;
$cmbC->config({ show_numbers => 0 });
printf "\tcmb->cmb_callback(%u, [%s], sum):\n", $count, $ilist;
our $num_calls = 0;
$res = $cmbC->cmb_callback($count, \@items, sub { $num_calls++; return 0 });
printf "\t\tnum_calls: %i\n", $num_calls;
printf "\t\tRESULT: %i\n", $res;

File: depend/libcmb/tests/test1.c
/*-
 * Copyright (c) 2018-2026 Devin Teske <dteske@FreeBSD.org>
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <cmb.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>

static int num_calls = 0;
static int
afunc(struct cmb_config *config, uint64_t seq, uint32_t nitems, char *items[])
{
	num_calls++;
	return (0);
}

int
main(void)
{
	uint32_t nitems = 4;
	int retval;
	uint64_t seq = 1;
	char *items[] = {"a", "b", "c", "d"};
	char itemstr[] = "a, b, c, d";
	static struct cmb_config config = {
		.options = CMB_OPT_NUMBERS,
		.delimiter = ",",
		.prefix = "\t[",
		.suffix = "]",
		.size_min = 2,
		.size_max = 3,
		.count = 6,
		.start = 4,
	};

	printf("config = {\n");
	printf("\t.options = 0x%08x,\n", config.options);
	printf("\t.delimiter = \"%s\",\n", config.delimiter);
	printf("\t.prefix = \"%s\",\n", config.prefix);
	printf("\t.suffix = \"%s\",\n", config.suffix);
	printf("\t.size_min = %u,\n", config.size_min);
	printf("\t.size_max = %u,\n", config.size_max);
	printf("\t.count = %"PRIu64",\n", config.count);
	printf("\t.start = %"PRIu64",\n", config.start);
	printf("}\n");

	printf("cmb_version(%u): %s\n", CMB_VERSION, cmb_version(CMB_VERSION));
	printf("cmb_version(%u): %s\n", CMB_VERSION_LONG,
	    cmb_version(CMB_VERSION_LONG));
	printf("size_min=%u size_max=%u\n", config.size_min, config.size_max);
	printf("cmb_count(config, %u) = %"PRIu64"\n", nitems,
	    cmb_count(&config, nitems));
	printf("cmb_print(config, %"PRIu64", %u, [%s]):\n", seq, nitems,
	    itemstr);
	retval = cmb_print(&config, seq, nitems, items);
	printf("\tRESULT: %i\n", retval);
	printf("cmb(config, %u, [%s]):\n", nitems, itemstr);
	printf("NOTE: { .start = %"PRIu64", .count = %"PRIu64" }\n",
	    config.start, config.count);
	retval = cmb(&config, nitems, items);
	printf("\tRESULT: %i\n", retval);

	/*
	 * Callbacks
	 */
	config.options = 0;
	config.action = afunc;
	printf("cmb_callback(config, %u, [%s], afunc):\n", nitems, itemstr);
	retval = cmb(&config, nitems, items);
	printf("\tnum_calls: %i\n", num_calls);
	printf("\tRESULT: %i\n", retval);

	return (EXIT_SUCCESS);
}
